### PR TITLE
write test for parser

### LINKED ISSUE #65: Write Tests for Parser
As a programmer I would like to apply my acquired knowledge about testing to the parser and thus cover the existing functionality of the parser with tests as much as possible.

---

## :white_check_mark: &nbsp; Conditions of satisfaction

- [ ] The functionality of the parser should be sufficiently _(pending code coverage)_ covered with tests.

### PR DESCRIPTION
**Pull request by kej-jay**
Fixes #65 

## Changes

*The functionality of the parser has not been ensured through tests*
- *it now is tested*

## Screenshots
grafik

## Checklist before merge

**Developer's responsibilities**
* [ ] Approved by at least one developer: @WolfgangDaniel @SanJSp 
* [ ] **Pull request build** has passed
* [ ] Production backend has been updated to reflect new features and fix bugs
* [X] tested locally (in all browsers if frontend)
* [ ] updated the docs
* [X] added tests where necessary

**Product Owner's responsibilities**
* [ ] No issues have been found on staging environment
* [ ] Product Owner has approved the release

### CODE REVIEW

File: client/src/utils/parser.test.js [hunk 1]
```diff
@@ -0,0 +1,351 @@
+const parser = require('./parser');
+const JSON_FOR_TESTING = {
```
Comment (kej-jay) [inline client/src/utils/parser.test.js:2]: @SanJSp  @WolfgangDaniel Please let me know ho you think about moving ths example json to a separate (helper) file and load it from there. I think that could be a good thing to do. Please let me know @WolfgangDaniel where that file could be located accourding to our structure, as I am a bit unsure about that

Comment (SanJSp) [reply client/src/utils/parser.test.js:2]: In my opinion really good thing to do, always outsource constants. Because this is a big constant, a new file is a great idea

File: client/src/utils/parser.js [hunk 2]
```diff
@@ -9,7 +9,7 @@ function generateCodeForMultiple(bpmnTasks) {
     let currentTask = bpmnTasks[task];
     let rpaTaskName = currentTask['_attributes']['arkRPA:task'];
     let rpaTaskInputString = currentTask['_attributes']['arkRPA:inputVars'];
-    let rpaTaskInput = rpaTaskInputString.replaceAll('/', '');
+    let rpaTaskInput = rpaTaskInputString.replace(/[\/]/g, '');
```
Comment (SanJSp) [inline client/src/utils/parser.js:12]: why was this change necessary?

File: client/src/utils/parser.test.js [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+                  "y": "330"
+                }
+              }
+            ]
+          }
+        ]
+      }
+    }
+  }
+};
```
Comment (SanJSp) [inline client/src/utils/parser.test.js:315]: add a newline please

File: client/src/utils/parser.test.js [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    expect(parserResultString.indexOf(BROWSER_ACTIVITY_NAME)).toBeLessThan(parserResultString.lastIndexOf(OPEN_BROWSER_CMD));
+    expect(parserResultString.indexOf(EXCEL_ACTIVITY_NAME)).toBeLessThan(parserResultString.lastIndexOf(OPEN_WORKBOOK_CMD));
+    expect(parserResultString.indexOf(EXCEL_ACTIVITY_NAME)).toBeLessThan(parserResultString.indexOf(FIND_EMPTY_ROW_CMD));
+
+    expect(parserResultString.indexOf(SETTING_STRING)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
+    expect(parserResultString.indexOf(LIBRARY_BROWSER)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
+    expect(parserResultString.indexOf(LIBRARY_EXCEL)).toBeLessThan(parserResultString.indexOf(LIBRARY_BROWSER));
+    expect(parserResultString.indexOf(LIBRARY_EXCEL)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
+  })
+})
```
Comment (SanJSp) [inline client/src/utils/parser.test.js:351]: newline at eof

File: client/src/utils/parser.test.js [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    expect(parserResultString).toMatch(OPEN_BROWSER_CMD);
+
+    expect(parserResultString.indexOf(BROWSER_ACTIVITY_NAME)).toBeLessThan(parserResultString.lastIndexOf(OPEN_BROWSER_CMD));
+    expect(parserResultString.indexOf(EXCEL_ACTIVITY_NAME)).toBeLessThan(parserResultString.lastIndexOf(OPEN_WORKBOOK_CMD));
+    expect(parserResultString.indexOf(EXCEL_ACTIVITY_NAME)).toBeLessThan(parserResultString.indexOf(FIND_EMPTY_ROW_CMD));
+
+    expect(parserResultString.indexOf(SETTING_STRING)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
+    expect(parserResultString.indexOf(LIBRARY_BROWSER)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
+    expect(parserResultString.indexOf(LIBRARY_EXCEL)).toBeLessThan(parserResultString.indexOf(LIBRARY_BROWSER));
+    expect(parserResultString.indexOf(LIBRARY_EXCEL)).toBeLessThan(parserResultString.indexOf(TASK_STRING));
```
Comment (SanJSp) [inline client/src/utils/parser.test.js:349]: normally I'd criticize that these are too many expect statements for the test. Seeing the upcoming change of the parser I will wave it through. A possible solution for this should be multiple tests (one test for containment and one for the order of the statements) or functions. Does outsourcing to functions work in tests?

**Review by SanJSp — COMMENTED** [review]
Just some code style and one comment for understanding. Looks fine, good work :)

File: client/src/utils/parser.js (hunk 2 above)
Comment (kej-jay) [reply client/src/utils/parser.js:12]: plain JS does not know pelaceAll, so this caused the tests to crash

**Review by SanJSp — APPROVED** [review]
